### DISCUSSION TITLE
latest/detection/core/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#1461] [Docs]
# latest/detection/core/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/latest/detection/core/

**Comment by jeffrymahbuubi** [comment]
Hi, I was wondering if the `sv.Detections.from_ultralytics` function would work with the `stream=True` argument, as the results will return a generator. I tried using `sv.Detections.from_ultralytics`, but it seems it can't process this generator. I wonder if there is any added benefit to using `stream=True` when combining Supervision with the Roboflow Inference library.

```python
def predict(self, image: np.ndarray) -> sv.Detections:
    results = self.model(image, verbose=False, device=self.device, stream=True)[0]
    return sv.Detections.from_ultralytics(results)
```

**Reply by LinasKo** [reply]
Hi @jeffrymahbuubi 👋

My understanding is that stream mode is typically used for videos and streams to avoid loading them into memory. We don't support `stream` in Ultralytics, but have a few alternatives.

For videos, you can use [get_video_frames_generator](https://supervision.roboflow.com/latest/utils/video/#supervision.utils.video.get_video_frames_generator). Our typical recommendation for streams is to use [InferencePipeline](https://inference.roboflow.com/using_inference/inference_pipeline/) with Roboflow models instead.

**Comment by olawalejuwonm** [comment]
This is a great library. I've used it for a research project, but I'm curious to know if there is any paper on it that i can cite?

**Reply by LinasKo** [reply]
Hi @olawalejuwonm 👋,

Glad to hear it's useful! If you go to the [Repository page](https://github.com/roboflow/supervision), on the right-hand side, you'll see a "Cite this repository" button. That should provide a copyable snippet for you.

<img width="365" alt="Screenshot 2024-08-20 at 4 31 19 PM" src="https://github.com/user-attachments/assets/346a4087-84fe-4958-a9e9-f513976da736">

Thanks again for the kind words!

**Comment by fakurten94** [comment]
Is there a way to get the mask XY coordinates from the output Detection object? As it outlines, you can only retrieve the XY coordinates of the bounding box but when using a segmentation model I wasn't quite sure if you could. If it can't it would be a really interesting feature to add as the YOLO models you can retrieve those coordinates.

**Reply by LinasKo** [reply]
Hi @fakurten94 👋 

Check out `sv.mask_to_polygons`. It converts each detected mask to a set of `xy` points. Is that what you were looking for?

**Comment by MrMerowinger** [comment]
Hello roboflow,
I have an issue when adding new key-value pairs to the data dictionary in sv.Detections.

```python
for image_path, image, annotations in detection_dataset_train:
    for xyxy, mask, confidence, class_id, tracker_id, data in annotations:
        data["category"] = ["dog"]
        data.__setitem__("position", [POSITION.CENTER])
        print(data)     
```

```
{'category': ['dog'], 'position': [<POSITION.CENTER: 'center'>]}
```

Its really odd that when I want to access it, the added values disapear.
```python
for image_path, image, annotations in detection_dataset_train:
    print(annotations.data)
```

```
{ 'category': [0], 'position': [0]}
```

What is reason for this and how can I add new pairs to the Detections after initialization? Thanks in advance!

I fixed the issue by saving the updated 'data' back to the annotations
```
annotations.data = data
```

**Reply by SkalskiP** [reply]
Hi, @MrMerowinger 👋🏻 Set your categories like this:

```python
annotations["category"] = ["dog", "cat", ...]  # The list should have the same number of elements as len(detections)
```

**Comment by 7ser23** [comment]
hi, I am lost in using Yolo for detection with the GPU. I have followed [Run a Model](https://inference.roboflow.com/quickstart/run_a_model/#install-inference). and made sure to run this line ```pip install inference-gpu ``` and I have checked that I am using T4 in Google Colab. I also run this: 
```
import torch
print(torch.cuda.is_available())  # True
print(torch.cuda.current_device())  # 0
print(torch.cuda.get_device_name(torch.cuda.current_device()))  # Tesla T4
```
and my model is loaded like this:
```
#model_id = "yolov8l-640"
model_id = "yolo11x-640"
model = get_model(model_id=model_id)
results = model.infer(frame, confidence=confidence, iou_threshold=iou, classes=classes)[0]
```
The GPU never goes above 0.1 in Google and I even get warnings that I am not using it. 
I want to process a video of 1 hour and in my m2 mac it takes around 54 minutes and in Google it takes even more time. Is there something else that I should be doing?. I am really lost, since when I trained a model in Google Colab it uses the GPU at 90% and I do the exact same verification as I just showed. Please help me

**Comment by MiiN1136** [comment]
Hi
I'm getting the following results from `GOOGLE_GEMINI_2_0` for an object detection task :
```
Here are the bounding box detections:
```json
[
  {"box_2d": [16, 300, 176, 352], "label": "person", "confidence": 1.0},
  {"box_2d": [16, 372, 162, 422], "label": "person", "confidence": 0.9},
  {"box_2d": [130, 548, 896, 840], "label": "person", "confidence": 1.0},
  {"box_2d": [160, 690, 392, 802], "label": "person", "confidence": 1.0},
  {"box_2d": [45, 0, 146, 46], "label": "person", "confidence": 0.9},
  {"box_2d": [34, 492, 322, 546], "label": "person", "confidence": 1.0}
]``` 
```
The issue I'm facing is that when I try to feed that `response.text` into `sv.Detections.from_vlm` using `sv.VLM.GOOGLE_GEMINI_2_0` as vlm, it parses them in a way that only `box_2d` and `label` are available for my final `Detections` object, whereas if I change the vlm to be `GOOGLE_GEMINI_2_5` the confidence key gets filled with correct results.
So, should I just continue generating output with Gemini 2.0 and parsing it with Gemini 2.5 ? or is there another approach ?
